python/src/nanoarrow/c_lib.py

```diff
@@ -120,15 +157,134 @@ def c_array(obj=None, requested_schema=None) -> CArray:
+) -> CArray:
+    """Create an ArrowArray wrapper from components
+
+    Given a schema, build an ArrowArray buffer-wise. This allows almost any array
+    to be assembled; however, requires some knowledge of the Arrow Columnar
+    specification. This function will do its best to validate the sizes and
+    content of buffers according to ``validation_level``, which can be set
+    to ``"full""`` for maximum safety.
+
+    Parameters
+    ----------
+
+    schema : schema-like
+        The data type of the desired array as sanitized by :func:`c_schema`.
+    length : int
+        The length of the output array.
+    buffers : Iterable of buffer-like or None
+        An iterable of buffers as sanitized by :func:`c_buffer`. Any object
+        supporting the Python Buffer protocol is accepted. Buffer data types
+        are not checked. A buffer value of ``None`` will skip setting a buffer
+        (i.e., that buffer will be of length zero and its pointer will
+        be ``NULL``).
+    null_count : int, optional
+        The number of null values, if known in advance. If -1 (the default),
+        the null count will be calculated based on the validity bitmap. If
+        the validity bitmap was set to ``None``, the calculated null count
+        will be zero.
+    offset : int, optional
+        The logical offset from the start of the array.
+    children : Iterable of array-like
+        An iterable of arrays used to set child fields of the array. Can contain
+        any object accepted by :func:`c_array`. Must contain the exact number of
+        required children as specifed by ``schema``.
+    validation_level: str, optional
+        One of "none" (no check), "minimal" (check buffer sizes that do not require
+        dereferencing buffer content), "default" (check all buffer sizes), or "full"
+        (check all buffer sizes and all buffer content).
+
+    Examples
+    --------
+
+    >>> import nanoarrow as na
+    >>> c_array = na.c_array_from_buffers(na.uint8(), 5, [None, b"12345"])
+    >>> na.c_array_view(c_array)
+    <nanoarrow.c_lib.CArrayView>
+    - storage_type: 'uint8'
+    - length: 5
+    - offset: 0
+    - null_count: 0
+    - buffers[2]:
+      - validity <bool[0 b] >
+      - data <uint8[5 b] 49 50 51 52 53>
+    - dictionary: NULL
+    - children[0]:
+    """
+    schema = c_schema(schema)
+    builder = CArrayBuilder.allocate()
+
+    # This is slightly wasteful: it will allocate arrays recursively and we are about
+    # to immediately release them and replace them with another value. We could also
+    # create an ArrowArrayView from the buffers, which would make it more
+    # straightforward to check the buffer types and avoid the extra structure
+    # allocation.
+    builder.init_from_schema(schema)
+
+    # Set buffers. This moves ownership of the buffers as well (i.e., the objects
+    # in the input buffers are replaced with an empty ArrowBuffer)
```

Comment:
This was a bit confusing (I ran into this while trying out the method):

```
In [74]: buf = na.c_buffer(b"0123")

In [75]: na.c_array_from_buffers(na.int32(), 1, [None, buf])
Out[75]: 
<nanoarrow.c_lib.CArray int32>
- length: 1
- offset: 0
- null_count: 0
- buffers: (0, 139688822562528)
- dictionary: NULL
- children[0]:

In [76]: na.c_array_from_buffers(na.int32(), 1, [None, buf])
---------------------------------------------------------------------------
NanoarrowException                        Traceback (most recent call last)
Cell In [76], line 1
----> 1 na.c_array_from_buffers(na.int32(), 1, [None, buf])

File ~/scipy/repos/arrow-nanoarrow/python/src/nanoarrow/c_lib.py:287, in c_array_from_buffers(schema, length, buffers, null_count, offset, children, validation_level)
    284 builder.resolve_null_count()
    286 # Validate + finish
--> 287 return builder.finish(validation_level=validation_level)

File src/nanoarrow/_lib.pyx:1818, in nanoarrow._lib.CArrayBuilder.finish()

File src/nanoarrow/_lib.pyx:422, in nanoarrow._lib.Error.raise_message_not_ok()

File src/nanoarrow/_lib.pyx:417, in nanoarrow._lib.Error.raise_message()

NanoarrowException: ArrowArrayFinishBuildingDefault() failed (22): Expected int32 array buffer 1 to have size >= 4 bytes but found buffer with 0 bytes

In [77]: buf
Out[77]: CBuffer(uint8[0 b] )
```

When passing python object through buffer protocol, the original object keeps owning the buffer content. We can't do something similar for CBuffer? (and that's because the CArray doesn't actually keep track of CBuffer objects, but only the C ArrowArray that has a pointer to the buffer?)

And where does this actually happen? I don't see anything in `CArrayBuilder.set_buffer` that would invalidate the buffer?

Reply:
> And where does this actually happen? I don't see anything in `CArrayBuilder.set_buffer` that would invalidate the buffer?

Ah, I suppose it is the `ArrowBufferMove` in the C code that does this?